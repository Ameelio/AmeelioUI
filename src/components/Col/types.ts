type NumberAttr =
  | number
  | '1'
  | '2'
  | '3'
  | '4'
  | '5'
  | '6'
  | '7'
  | '8'
  | '9'
  | '10'
  | '11'
  | '12';
type ColSize = true | 'auto' | NumberAttr;
export type ColSpec =
  | ColSize
  | { span?: ColSize; offset?: NumberAttr; order?: NumberAttr };