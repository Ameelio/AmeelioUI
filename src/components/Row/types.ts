export type RowSpec = number | { cols: number }