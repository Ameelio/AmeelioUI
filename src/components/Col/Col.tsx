import React, { Component } from 'react';
import BootstrapCol, { ColProps as BootstrapColProps} from 'react-bootstrap/Col';
import { ElementType } from 'react';
import { ColSpec } from './types';

export interface ColProps extends BootstrapColProps {
    /**
     * Custom element type for this component
     * @default <div>
     */
    as?: ElementType;

    /**
     * The number of columns that will fit next to each other on large devices (>=992px)
     */
    lg?: ColSpec;

    /**
     * The number of columns that will fit next to each other on medium devices (>=768px)
     */
    md?: ColSpec;

    /**
     * The number of columns that will fit next to each other on small devices (>=576px)
     */
    sm?: ColSpec;

    /**
     * The number of columns that will fit next to each other on extra large devices (>=1200px)
     */
    xl?: ColSpec;

    /**
     * The number of columns that will fit next to each other on extra small devices (>=576px)
     */
    xs?: ColSpec;

    /**
     * Change the underlying component CSS base class name and modifier class names prefix. This is an escape hatch for working with heavily customized bootstrap css
     * @default 'row'
     */
    bsPrefix?: string;

    /**
     * One or more class names to be added to the root element of this component
     */
    className?: string;
}

export default class Col extends Component<ColProps> {
    public render() {
        const className = `y-col ${this.props.className ? this.props.className : ''}`
        return (
            <BootstrapCol {...this.props} className={className}/>
        );
    }
}
