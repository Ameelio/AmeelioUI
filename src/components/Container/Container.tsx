import React, { Component } from 'react';
import BootstrapContainer, { ContainerProps as BootstrapContainerProps} from 'react-bootstrap/Container';
import { ElementType } from 'react';
import { ContainerSize } from './types';

export { ContainerSize }

export interface ContainerProps extends BootstrapContainerProps {
    /**
     * Custom element type for this component
     * @default <div>
     */
    as?: ElementType;

    /**
     * Allow the Container to fill all of its available horizontal space
     * @default false
     */
    fluid?: boolean | ContainerSize;

    /**
     * Change the underlying component CSS base class name and modifier class names prefix. This is an escape hatch for working with heavily customized bootstrap css
     * @default 'container'
     */
    bsPrefix?: string;

    /**
     * One or more class names to be added to the root element of this component
     */
    className?: string;
}

export default class Container extends Component<ContainerProps> {
    public render() {
        const className = `y-container ${this.props.className ? this.props.className : ''}`
        return (
            <BootstrapContainer {...this.props} className={className}/>
        );
    }
}
