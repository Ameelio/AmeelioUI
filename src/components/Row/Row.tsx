import React, { Component } from 'react';
import BootstrapRow, { RowProps as BootstrapRowProps} from 'react-bootstrap/Row';
import { ElementType } from 'react';
import { RowSpec } from './types';

export interface RowProps extends BootstrapRowProps {
    /**
     * Custom element type for this component
     * @default <div>
     */
    as?: ElementType;

    /**
     * The number of columns that will fit next to each other on large devices (>=992px)
     */
    lg?: RowSpec;

    /**
     * The number of columns that will fit next to each other on medium devices (>=768px)
     */
    md?: RowSpec;

    /**
     * Removes the gutter spacing between Cols as well as any added negative margins
     * @default false
     */
    noGutters?: boolean;

    /**
     * The number of columns that will fit next to each other on small devices (>=576px)
     */
    sm?: RowSpec;

    /**
     * The number of columns that will fit next to each other on extra large devices (>=1200px)
     */
    xl?: RowSpec;

    /**
     * The number of columns that will fit next to each other on extra small devices (>=576px)
     */
    xs?: RowSpec;

    /**
     * Change the underlying component CSS base class name and modifier class names prefix. This is an escape hatch for working with heavily customized bootstrap css
     * @default 'row'
     */
    bsPrefix?: string;

    /**
     * One or more class names to be added to the root element of this component
     */
    className?: string;
}

export default class Row extends Component<RowProps> {
    public render() {
        const className = `y-row ${this.props.className ? this.props.className : ''}`
        return (
            <BootstrapRow {...this.props} className={className}/>
        );
    }
}
