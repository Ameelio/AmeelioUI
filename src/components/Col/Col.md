### Examples

#### Auto-layout columns
`Col` will render columns with equal width if no column width is provided.

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className='border border-primary'>1 of 2</Col>
    <Col className='border border-primary'>2 of 2</Col>
  </Row>
  <Row>
    <Col className='border border-primary'>1 of 3</Col>
    <Col className='border border-primary'>2 of 3</Col>
    <Col className='border border-primary'>3 of 3</Col>
  </Row>
</Container>
```

#### One column width
If you set the width of only one column, the other columns will automatically resize around it. 

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col className='border border-primary'>1 of 3</Col>
    <Col xs={6} className='border border-primary'>2 of 3 (wider)</Col>
    <Col className='border border-primary'>3 of 3</Col>
  </Row>
  <Row>
    <Col className='border border-primary'>1 of 3</Col>
    <Col xs={5} className='border border-primary'>2 of 3 (wider)</Col>
    <Col className='border border-primary'>3 of 3</Col>
  </Row>
</Container>
```

#### Variable width content
Setting the column value to `"auto"` will size columns based on the width of their content.

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row  className="justify-content-md-center">
    <Col xs lg='2' className='border border-primary'>1 of 3</Col>
    <Col md='auto' className='border border-primary'>Variable width content</Col>
    <Col xs lg='2' className='border border-primary'>3 of 3</Col>
  </Row>
  <Row>
    <Col className='border border-primary'>1 of 3</Col>
    <Col md='auto' className='border border-primary'>Variable width content</Col>
    <Col xs lg='2' className='border border-primary'>3 of 3</Col>
  </Row>
</Container>
```

#### Responsive grids
`Col` allows column widths to be specified for each breakpoint size (xs, sm, md, lg, xl). For each breakpoint, the number of colums to span can ben set, or you can simply set the prop to `true` for automatic widths.

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col sm={8} className='border border-primary'>sm=8</Col>
    <Col sm={4} className='border border-primary'>sm=4</Col>
  </Row>
  <Row>
    <Col sm className='border border-primary'>sm=true</Col>
    <Col sm className='border border-primary'>sm=true</Col>
    <Col sm className='border border-primary'>sm=true</Col>
  </Row>
</Container>
```

Breakpoints can also be mixed and matched to create different grids based on screen size.

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  {/* Stack columns on mobile by making one full-width and the other half-width */}
  <Row>
    <Col xs={12} md={8} className='border border-primary'>xs=12 md=8</Col>
    <Col xs={6} md={4} className='border border-primary'>xs=6 md=4</Col>
  </Row>

  {/* Columns start at 50% width on mobile and bump up to 33.3% width on desktop */}  
  <Row>
    <Col xs={6} md={4} className='border border-primary'>xs=6 md=4</Col>
    <Col xs={6} md={4} className='border border-primary'>xs=6 md=4</Col>
    <Col xs={6} md={4} className='border border-primary'>xs=6 md=4</Col>
  </Row>

  {/* Columns always have 50% width on mobile and desktop */}  
  <Row>
    <Col xs={6} className='border border-primary'>xs=6</Col>
    <Col xs={6} className='border border-primary'>xs=6</Col>
  </Row>
</Container>
```

`Col` breakpoint props can also be specified in `object` form: `{span: number, order:number, offset: number}` for specifying offsets and visual ordering.

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col xs className='border border-primary'>First, but unordered</Col>
    <Col xs={{ order: 12 }} className='border border-primary'>Second, but last</Col>
    <Col xs={{ order: 1 }} className='border border-primary'>Third, but second</Col>
  </Row>
</Container>
```

The `order` also supports `first` and `last`

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col xs={{order: 'last' }} className='border border-primary'>First, but last</Col>
    <Col xs className='border border-primary'>Second, but unordered</Col>
    <Col xs={{ order: 'first' }} className='border border-primary'>Third, but first</Col>
  </Row>
</Container>
```

For offsetting columns, you can set an `offset` value. For more general layout, use the margin utility class.

```js
import Container from '../Container';
import Row from '../Row';

<Container>
  <Row>
    <Col md={4} className='border border-primary'>md=4</Col>
    <Col md={{ span: 4, offset: 4 }} className='border border-primary'>{`md={{ span: 4, offset: 4 }}`}</Col>
  </Row>
  <Row>
    <Col md={{ span: 3, offset: 3 }} className='border border-primary'>{`md={{ span: 3, offset: 3 }}`}</Col>
    <Col md={{ span: 3, offset: 3 }} className='border border-primary'>{`md={{ span: 3, offset: 3 }}`}</Col>
  </Row>
  <Row>
    <Col md={{ span: 6, offset: 3 }} className='border border-primary'>{`md={{ span: 6, offset: 3 }}`}</Col>
  </Row>
</Container>
```
