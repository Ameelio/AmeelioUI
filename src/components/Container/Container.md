### Examples

#### Basic container
`Container` provides a means to center and horizontally pad content. Containers provide responsive pixel width.

```js
import Row from '../Row';
import Col from '../Col';

<Container className='border border-primary'>
  <Row>
    <Col>1 of 1</Col>
  </Row>
</Container>
```

#### Fluid container
Fluid containers will span the full width on all viewports and device sizes.

```js
import Row from '../Row';
import Col from '../Col';

<Container fluid className='border border-primary'>
  <Row>
    <Col>1 of 1</Col>
  </Row>
</Container>
```

#### Fluid contianer with breakpoint
Breakpoints can be combined with the `fluid` prop to set the container as fluid up to a specified breakpoint.

```js
import Row from '../Row';
import Col from '../Col';

<Container fluid="md" className='border border-primary'>
  <Row>
    <Col>1 of 1</Col>
  </Row>
</Container>
```
