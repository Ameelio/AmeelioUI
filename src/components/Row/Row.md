### Examples

#### Setting column widths
`Row` allows you to specify the number of columns that will fit next to each other for each breakpoint size.

```js
import Container from '../Container';
import Col from '../Col';

<Container>
  <Row xs={2} md={4} lg={6}>
    <Col className='border border-primary'>1 of 2</Col>
    <Col className='border border-primary'>2 of 2</Col>
  </Row>
  <Row xs={1} md={2}>
    <Col className='border border-primary'>1 of 3</Col>
    <Col className='border border-primary'>2 of 3</Col>
    <Col className='border border-primary'>3 of 3</Col>
  </Row>
</Container>
```

Note that `Row` column widths will override Col widths set on lower breakpoints when viewed on larger screens. The elow `Col` size will be overriden by the `Row` size on medium and larger screens.

```js
import Container from '../Container';
import Col from '../Col';

<Container>
  <Row md={4}>
    <Col className='border border-primary'>1 of 3</Col>
    <Col xs={6} className='border border-primary'>2 of 3</Col>
    <Col className='border border-primary'>3 of 3</Col>
  </Row>
</Container>
```
