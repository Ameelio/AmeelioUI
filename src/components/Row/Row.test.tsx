import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Row, { RowProps } from './Row';
import Container from '../Container';
import Col from '../Col';

const sampleText = 'Some text';

describe('<Row />', () => {
    let component: ShallowWrapper<RowProps>;

    describe('with default options', () => {
        beforeEach(() => {
            component = shallow(<Container>
                <Row>
                    <Col>{sampleText}</Col>
                </Row>
            </Container>);
        });

        it('matches its snapshot', () => {
            expect(component).toMatchSnapshot();
        });
    });

    describe('with row breakpoints', () => {
        beforeEach(() => {
            component = shallow(<Container fluid>
                <Row xs={2} md={4} lg={6}>
                    <Col>{sampleText}</Col>
                    <Col>{sampleText}</Col>
                </Row>
            </Container>);
        });

        it('matches its snapshot', () => {
            expect(component).toMatchSnapshot();
        });
    });

    describe('with row breakpoints that override column breakpoints', () => {
        beforeEach(() => {
            component = shallow(<Container fluid>
                <Row md={4}>
                    <Col>{sampleText}</Col>
                    <Col xs={6}>{sampleText}</Col>
                    <Col>{sampleText}</Col>
                </Row>
            </Container>);
        });

        it('matches its snapshot', () => {
            expect(component).toMatchSnapshot();
        });
    });

    describe('with custom base class prefix', () => {
        beforeEach(() => {
            component = shallow(<Container>
                <Row bsPrefix="ameelio-container">
                    <Col>{sampleText}</Col>
                </Row>
            </Container>);
        });

        it('matches its snapshot', () => {
            expect(component).toMatchSnapshot();
        });
    });

    describe('with additional className', () => {
        beforeEach(() => {
            component = shallow(<Container>
                <Row className="TEST_CLASSNAME">
                    <Col>{sampleText}</Col>
                </Row>
            </Container>);
        });

        it('matches its snapshot', () => {
            expect(component).toMatchSnapshot();
        });
    });
});