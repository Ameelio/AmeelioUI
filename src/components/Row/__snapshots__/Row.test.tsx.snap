// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> with additional className matches its snapshot 1`] = `
<Container
  className="y-container "
  fluid={false}
>
  <Row
    className="TEST_CLASSNAME"
  >
    <Col>
      Some text
    </Col>
  </Row>
</Container>
`;

exports[`<Row /> with custom base class prefix matches its snapshot 1`] = `
<Container
  className="y-container "
  fluid={false}
>
  <Row
    bsPrefix="ameelio-container"
  >
    <Col>
      Some text
    </Col>
  </Row>
</Container>
`;

exports[`<Row /> with default options matches its snapshot 1`] = `
<Container
  className="y-container "
  fluid={false}
>
  <Row>
    <Col>
      Some text
    </Col>
  </Row>
</Container>
`;

exports[`<Row /> with row breakpoints matches its snapshot 1`] = `
<Container
  className="y-container "
  fluid={true}
>
  <Row
    lg={6}
    md={4}
    xs={2}
  >
    <Col>
      Some text
    </Col>
    <Col>
      Some text
    </Col>
  </Row>
</Container>
`;

exports[`<Row /> with row breakpoints that override column breakpoints matches its snapshot 1`] = `
<Container
  className="y-container "
  fluid={true}
>
  <Row
    md={4}
  >
    <Col>
      Some text
    </Col>
    <Col
      xs={6}
    >
      Some text
    </Col>
    <Col>
      Some text
    </Col>
  </Row>
</Container>
`;
